### DISCUSSION TITLE
Menu does not show up

### DISCUSSION THREAD

**Discussion by CoolHandDevil505** [BazzaCuda/CustomMenu#12] [Q&A] [answered]
Hi,

I’m using W11, the icon in the taskbar is present, but right clicking on desktop brings the default win menu…
Shift + right click does not close CustomMenu.

Maybe a software incompatibility? How to tell?

Thank you.

**Comment by BazzaCuda** [comment]
Hi, that's very odd.
Which build of Windows11 are you running, please? (System Settings / System  / About / Windows specifications / OS build)
Are you up to date with Win11 updates?

I have CustomMenu running in a Win11 VM (downloaded from Microsoft), build 22621.1702, so it _should_ work in any later build.

**Comment by CoolHandDevil505** [comment]
Hi,

Windows 11 22H2 build 22621.2215, up to date.

I guess it is something related to my system… I may try to uninstall conflicting software, but I don't know where to start…
Do you know a way to diagnose the problem?

**Comment by BazzaCuda** [comment]
I'm a bit stuck at the moment trying to second-guess what could be different about your Win11.
I've released a small tweak to the software (v1.0.4) to simplify how it determines if a Ctrl key is down when you right-click the desktop. If it thought (either rightly or wrongly) that a Ctrl key was down, it definitely wouldn't do anything, and would therefore exhibit the behavior that you describe.
Having said that, I can't replicate your problem on my own Win11 VM; it works fine.

In the release I've asked others to report if they have the same issue as you do. But until we have more info, it's like trying to find a needle in a haystack. Do you have other Windows computers you can run CustomMenu on to see if you can spot any possible variations between those and your Win11 machine?

Edit: p.s. I've just updated my VM to the exact same build as yours; still no problems.

**Reply by CoolHandDevil505** [reply]
Thank you for the quick reply.
The 1.0.4 behave the same way for me unfortunately. I have a W10 computer, I will try asap, do you think it can help?
Another thing to mention, once launched, I cannot close the program, by shift+right click on desktop, by right clicking on the systray nor the task manager (i see an entry for CustomMenu, but quickly disappearing after opening of the task manager window)

**Reply by CoolHandDevil505** [reply]
Well… I have the same issue on my W10 machine (21H2 19044.2965).

**Reply by BazzaCuda** [reply]
Some questions (as I think something else might be interfering with the process, going by what you said about Task Manager).

What anti-virus software do you have running, and is that the same on both machines?

Also, when you hover the mouse over the system tray icon you should get a hint like this:
![hover](https://github.com/BazzaCuda/CustomMenu/assets/22550919/44b01be9-f14c-46d7-b98d-16838200accd)
Does that happen?


Then when you right-click the system tray icon you should get a single menu item:
![right-click](https://github.com/BazzaCuda/CustomMenu/assets/22550919/e0530e88-71f8-4697-81fb-551cf6bcc2b1)
Does that menu item display correctly?

Finally, if you [left-] click the menu item you need to be careful that your mouse pointer is over the system tray and not over any part of the desktop (a pain, I know), otherwise it's treated as a desktop click rather than a click on the menu item.

**Reply by CoolHandDevil505** [reply]
I am running Microsoft Defender on both machines.

![Capture d'écran 2023-08-24 002402](https://github.com/BazzaCuda/CustomMenu/assets/37642912/915d66fe-f8e3-4956-ab77-87c13c314ff4)
![Capture d'écran 2023-08-24 002423](https://github.com/BazzaCuda/CustomMenu/assets/37642912/0edcd449-f167-4722-94b1-7261cf0994e0)
And when I click the menu, the pointer is over the systray.

**Reply by BazzaCuda** [reply]
Interesting. I had a feeling you were going to say Microsoft Defender. I would suggest you either temporarily disable MD, or add CustomMenu.exe to Defender's list of exceptions so that it ignores it....

Settings > Update & Security > Windows Security > Virus & threat protection > Virus & threat protection settings > Manage settings > Exclusions > Add or remove exclusions > Add Exclusion > File
Then browse to and select CustomMenu.exe

I did find, yesterday, on my Win11 VM (which has Microsoft Defender installed by default) that submenus display quicker with MD disabled, so I know that it does interfere with the smooth running of the process but without reporting it as a problem to user! 

Try that (or disable it, temporarily) and see if there's any improvement.

At what point does CustomMenu disappear from Task Manager? (And when it does so, if you hover over the tray icon, the icon should disappear (?))

**Reply by CoolHandDevil505** [reply]
Hi,

Tried to deactivate MD (or add exclusion), tried to remove all software running at startup, tried safe mode… with no luck.
Maybe some dependency not present on my system for some reason?

I found however, that CustomMenu is actually present in task manager, but in the details section, not in the process (where I see it for half a second). And when I close it from there, the systray icon close itself.

**Reply by BazzaCuda** [reply]
All very odd. I have to be honest, I'm a bit stumped at the moment. My gut feeling is that Microsoft Defender is interfering with the process somehow, but how and why, I'm not sure. Also, no one else so far has reported the same issue, and there must be lots of people running MD. Without sitting at your computer(s) myself and investigating I'm at a loss as to what to do. I suppose I could give you a debug version and let it try to display messages to you when you click the desktop, if you're interested in trying that approach?

**Reply by CoolHandDevil505** [reply]
Oh yes! I will be happy to try that. Also I never looked at Pascal but, is there a dev software with a step execution type that I can use?

**Reply by BazzaCuda** [reply]
CustomMenu was developed using Embarcadero Delphi/RAD Studio Community Edition, with some additional component libraries, e.g. Virtual Shell. You're welcome to download everything and attempt to debug it, although there's quite a lot to configure. However, because it intercepts Windows messages in a hook, stepping through the code is difficult as the hook interferes with the Delphi IDE. I've even included a note to that effect in the source code as a warning to other developers. Trying to debug the hook itself is easier when you output messages from it to determine what's going on. Especially if you use my own BazDebugWindow app (a modification of the GExperts Debug Window). I probably won't be able to look at this again until Sunday, but I'll have a think and work out what I can send you so we can work out which bits of the code are being affected.

**Reply by CoolHandDevil505** [reply]
I just remember that I had Nilesoft Shell installed, I removed it, but no change…

**Reply by BazzaCuda** [reply]
LOL! I don't know if that would have affected CustomMenu, but Nilesoft is the sort of software that might.

OK. First you need to download my BazDebugWindow app, which we're going to try and send messages to.
https://github.com/BazzaCuda/BazDebugWindow/files/12181796/BazDebugWindow_release_v1_0_3.zip

Just unzip it to any folder and run BazDebugWindow.exe
It will sit in your system tray. You should be able to double-click the icon to force it to display its window.
You can close the window again and the icon will remain.

Then run this app:
[testDebugWindow.zip](https://github.com/BazzaCuda/CustomMenu/files/12448013/testDebugWindow.zip)

All it does is to send a message, "Hello from testDebugWindow" to BazDebugWindow.
If the BazDebugWindow window isn't open, it will open it.
Then it will give you a "Done" message for you to OK.

Let me know how you get on.

p.s. You should be able to right-click the icon in the system tray and choose "Shutdown".
If you then run testDebugWindow again, BazDebugWindow should launch all by itself.

**Reply by CoolHandDevil505** [reply]
![Capture d'écran 2023-08-27 223919](https://github.com/BazzaCuda/CustomMenu/assets/37642912/fc2daa19-6aa4-40c3-bdf1-e7b36b3ea5ed)
Works as described, also with BazDebugWindow running itself (MD was quite worried by testDebugWindow fyi).

**Reply by BazzaCuda** [reply]
I'm not surprised about MD. It seems to dislike any executable which isn't digitally signed, rather than just the ones which actually pose a legitimate threat. That's one of the [many] reasons I've disabled it for years. As a developer creating executables all day every day, MD became a major pain. I use MalwareBytes instead; far superior.

OK. Next step.
Run this version of CustomMenu instead.
[CustomMenu.zip](https://github.com/BazzaCuda/CustomMenu/files/12454226/CustomMenu.zip)

As you move the mouse around your desktop, you should get loads of repeated "message hook" messages.
If you click the pause button (fourth button along) and scroll to the very bottom of the list, there should be three non-zero numbers. They will be different to those shown here but that's ok.
![Clipboard Image](https://github.com/BazzaCuda/CustomMenu/assets/22550919/2ddd070b-7bae-4978-baff-a05f35a643e6)

**Reply by CoolHandDevil505** [reply]
I get : hDesktop = 0, hDefView = 0, hProgman = 131300

**Reply by BazzaCuda** [reply]
Well! That explains that then, doesn't it :P

Are you running any desktop-/ProgramManager-replacement software?
Also, do you have a Windows slideshow running on your desktop, so Windows changes your desktop image at intervals?

**Reply by CoolHandDevil505** [reply]
I am using vanilla desktop from W11, not sure what you mean by program manager though… Sorry…

**Reply by BazzaCuda** [reply]
Try running this and tell me what numbers you get in the debug window:
[CustomMenu.zip](https://github.com/BazzaCuda/CustomMenu/files/12465850/CustomMenu.zip)

I've removed the "message hook" message entirely. We don't need it. I think I've identified what the problem might be.

**Reply by CoolHandDevil505** [reply]
2	hDesktop = 0
2	hWorkerW = 0
2	hDefView = 0
2	hProgman = 65780

**Reply by BazzaCuda** [reply]
Try this one please. 
[CustomMenu.zip](https://github.com/BazzaCuda/CustomMenu/files/12466396/CustomMenu.zip)

It now tries 3 different ways to get Windows to divulge the handle of its DefView window :D

**Reply by CoolHandDevil505** [reply]
2	hDesktop = 65792
2	hWorkerW = 0
2	hDefView = 131324
2	hProgman = 65780

**Reply by BazzaCuda** [reply]
Try right-clicking the desktop! ;)

**Reply by CoolHandDevil505** [reply]
![Capture d'écran 2023-08-29 182541](https://github.com/BazzaCuda/CustomMenu/assets/37642912/9a0452de-9f74-4a89-a2f5-848a370b3c43)
Right click

![Capture d'écran 2023-08-29 182555](https://github.com/BazzaCuda/CustomMenu/assets/37642912/af1d4c58-6803-4787-b537-53aafded3032)
Shift + Right click (normal?)

2	hDesktop = 65776
2	hWorkerW = 0
2	hDefView = 65772
2	hProgman = 131304

**Reply by BazzaCuda** [reply]
So CustomMenu is now working?
The first screenshot looks like the standard Windows menu and the second screenshot looks like CustomMenu, yes?

**Reply by CoolHandDevil505** [reply]
Unfortunately it's not… right-clicking brings w11 menu, and shift right click brings w10 style menu…

**Reply by CoolHandDevil505** [reply]
It is not working on my w10 either.

**Reply by BazzaCuda** [reply]
This is so bizarre!
Do you have all the same software running on both machines?
I can't think what might be running that could interfere so much with the desktop like this.
It seems to be finding the desktop window now, so what's its problem? :D

**Reply by CoolHandDevil505** [reply]
More or less, it is the same software on both, yes… It is my old prod machine, so it is pretty much the same. I did not use any migration software, it is a brand new w11 on my new pc, but I must have some kind of troublemaker for CustomMenu. I'm so sorry, I wish I could be more helpful…

**Reply by BazzaCuda** [reply]
More numbers.
[CustomMenu.zip](https://github.com/BazzaCuda/CustomMenu/files/12467388/CustomMenu.zip)

If you run this you should get a "Mouse Test" message in the debug window.
Make a note of the hDesktop number.
Then left-click and right-click on the desktop several times.
Every time you click on the desktop with either mouse button, a debug message will tell you the window number that received the mouse click.
This should correspond to the hDesktop number that it first reported.

**Reply by CoolHandDevil505** [reply]
mouseWnd (65772) does not match hDesktop (65776) when left or right clicking on desktop.

**Reply by BazzaCuda** [reply]
Well, that's the issue then. The question is, why don't they match?
Why isn't Windows reporting the correct window handle?

Can you send me a screenshot of the first lot of debug messages that it reported please, starting with "Mouse Test".

**Reply by BazzaCuda** [reply]
Last attempt for today.
[CustomMenu.zip](https://github.com/BazzaCuda/CustomMenu/files/12468055/CustomMenu.zip)

At least I now know what the problem is.
If this version doesn't fix it, I have some more ideas we can try tomorrow.
I'm determined to get to the bottom of this for you.

**Comment by CoolHandDevil505** [comment]
> Well, that's the issue then. The question is, why don't they match? Why isn't Windows reporting the correct window handle?
> 
> Can you send me a screenshot of the first lot of debug messages that it reported please, starting with "Mouse Test".

![Capture d'écran 2023-08-29 210857](https://github.com/BazzaCuda/CustomMenu/assets/37642912/09e4f119-c026-4c0c-b687-a6b202b08733)

**Comment by CoolHandDevil505** [comment]
> Last attempt for today. [CustomMenu.zip](https://github.com/BazzaCuda/CustomMenu/files/12468055/CustomMenu.zip)
> 
> At least I now know what the problem is. If this version doesn't fix it, I have some more ideas we can try tomorrow. I'm determined to get to the bottom of this for you.

![Capture d'écran 2023-08-29 211346](https://github.com/BazzaCuda/CustomMenu/assets/37642912/ba1e1025-2687-4878-a5bf-067296a9ce8d)

**Reply by BazzaCuda** [reply]
Good morning!
Today is the day, I reckon :D

Ok. Two steps. On your Windows11 machine.

Step One. 
Fire-up a new instance of CustomMenu, click the desktop a few times and then send me a screenshot of the first few messages from the debug window.

Step two.
Run this program:
[WinDowse.zip](https://github.com/BazzaCuda/CustomMenu/files/12473412/WinDowse.zip)
This will tell us the actual numbers Windows should be reporting to us.

Go to the Options tab.
![Clipboard Image](https://github.com/BazzaCuda/CustomMenu/assets/22550919/86f4be4d-c14a-4482-9aae-a907972e730a)
Check that a) Integer Format is set to Decimal, and b) the start/stop hotkey is set to CAPSLOCK. You can change this to any key you prefer, but allow for that change when reading the following instructions.

Go to the Window tab
![Clipboard Image (1)](https://github.com/BazzaCuda/CustomMenu/assets/22550919/59e3ef3d-9638-4abb-85c0-88123e010e7c)
Notice the "Dowse" button. When you press the CAPSLOCK key, this will change to "Stop". Press CAPSLOCK again and this will revert to "Dowse"

1. Press the CAPSLOCK key so that the button says "Stop".
2. Move your mouse so that the pointer is over a completely empty part of your desktop and leave it there.
3. The "Text" box should change to something like "FolderView".
4. Leave the mouse where it is and press CAPSLOCK again so that the button says "Dowse".
5. Click the "Save" button and save all the Windowse info to a text file (e.g. on your desktop).
6. Drag and drop the text file into a reply to me here.

Please don't reboot at any time or you'll have to do this all again as all the numbers will change.
Also, I might send you a special version of CustomMenu which is set to only use your numbers.

Thanks in advance.

**Reply by CoolHandDevil505** [reply]
Step one
![Capture d'écran 2023-08-30 111923](https://github.com/BazzaCuda/CustomMenu/assets/37642912/a7da20a9-deac-46f9-823c-6625ea376049)

Step two
[WinDowse.txt](https://github.com/BazzaCuda/CustomMenu/files/12474091/WinDowse.txt)

**Reply by BazzaCuda** [reply, marked as answer]
Try this one.
[CustomMenu.zip](https://github.com/BazzaCuda/CustomMenu/files/12474496/CustomMenu.zip)

**Reply by CoolHandDevil505** [reply]
![Capture d'écran 2023-08-30 160121](https://github.com/BazzaCuda/CustomMenu/assets/37642912/34c0b711-5c75-4853-be70-9c7364d58530)
 It is working! How did you do it?

**Reply by BazzaCuda** [reply]
YAY!!!

Under certain circumstances, Windows creates completely different invisible windows for the desktop than it normally does.
One of those circumstances, for example, is if people set Windows to periodically change the desktop wallpaper.
I'm sure, like in your case, there must be others.

When that happens, Windows creates invisible "WorkerW" windows - you can see them in the text file you sent me.
That's ok, but Windows also changes which window gets all the mouse movement/click messages. Thanks, Microsoft! :P
For years (since Windows95 ?), it's been the desktop window that you can see which gets the mouse messages; a window called "FolderView".
Under these other circumstances, it's a window called SHELLDLL_DefView which is the invisible parent of your desktop but is now owned by one(!) of the WorkerW windows.
Utterly ridiculous!

I've simply changed CustomMenu so that when it detects mouse messages it knows that if they're being created for either the actual desktop window or for this other DefView window, to treat them as being desktop mouse messages.

Thank you very much for your help and for your patience. I couldn't have done this without you as neither my Windows10 nor my Windows 11 VM create those "WorkerW" windows. I knew nothing about them until yesterday.

I will create an official release which fixes the problem but doesn't have all the debug messages in it.
If you shutdown BazDebugWindow.exe and delete the BazDebugWindow.exe file, or if you rename it to something invalid like BazDebugWindow.ex_, the messages that CustomMenu is currently sending won't go anywhere and it will just operate as normal, although possibly a little slower.

If you have any other issues, please let me know immediately. I'm not 100% sure yet whether my fix currently works in all circumstances where those WorkerW windows are concerned. I will have to investigate them further.

Did you also try it out on your Windows10 machine?

Kind regards,
Baz.

**Reply by CoolHandDevil505** [reply]
Yes I did try on W10, and it works too. And yes on both of them, I have random wallpapers 😏
So it wasn't me 😁
I was happy to help, and will be in the future…
That was one nasty behavior to find as I understand, but debugging is always fun, hey?
Cheers.
